polkadot/node/network/availability-recovery/src/task.rs

```diff
@@ -0,0 +1,851 @@
+			// or have gotten firm 'No' responses from enough validators.
+			if can_conclude(
+				validators.len(),
+				self.requesting_chunks.total_len(),
+				self.chunk_count(),
+				params,
+				error_count,
+			) {
+				gum::debug!(
+					target: LOG_TARGET,
+					candidate_hash = ?params.candidate_hash,
+					received_chunks_count = ?self.chunk_count(),
+					requested_chunks_count = ?self.requesting_chunks.len(),
+					threshold = ?params.threshold,
+					"Can conclude availability for a candidate",
+				);
+				break
+			}
+		}
+
+		(total_received_responses, error_count)
+	}
+}
+
+/// A stateful reconstruction of availability data in reference to
+/// a candidate hash.
+pub struct RecoveryTask<Sender: overseer::AvailabilityRecoverySenderTrait> {
+	sender: Sender,
+	params: RecoveryParams,
+	strategies: VecDeque<Box<dyn RecoveryStrategy<Sender>>>,
+	state: State,
+}
+
+impl<Sender> RecoveryTask<Sender>
+where
+	Sender: overseer::AvailabilityRecoverySenderTrait,
+{
+	/// Instantiate a new recovery task.
+	pub fn new(
+		sender: Sender,
+		params: RecoveryParams,
+		strategies: VecDeque<Box<dyn RecoveryStrategy<Sender>>>,
+	) -> Self {
+		Self { sender, params, strategies, state: State::new() }
+	}
+
+	async fn in_availability_store(&mut self) -> Option<AvailableData> {
+		if !self.params.bypass_availability_store {
+			let (tx, rx) = oneshot::channel();
+			self.sender
+				.send_message(AvailabilityStoreMessage::QueryAvailableData(
+					self.params.candidate_hash,
+					tx,
+				))
+				.await;
+
+			match rx.await {
+				Ok(Some(data)) => return Some(data),
+				Ok(None) => {},
+				Err(oneshot::Canceled) => {
+					gum::warn!(
+						target: LOG_TARGET,
+						candidate_hash = ?self.params.candidate_hash,
+						"Failed to reach the availability store",
+					)
+				},
+			}
+		}
+
+		None
+	}
+
+	/// Run this recovery task to completion. It will loop through the configured strategies
+	/// in-order and return whenever the first one recovers the full `AvailableData`.
+	pub async fn run(mut self) -> Result<AvailableData, RecoveryError> {
+		if let Some(data) = self.in_availability_store().await {
+			return Ok(data)
+		}
+
+		self.params.metrics.on_recovery_started();
+
+		let _timer = self.params.metrics.time_full_recovery();
+
+		let res = loop {
+			if let Some(mut current_strategy) = self.strategies.pop_front() {
+				// Make sure we are not referencing futures from past RecoveryStrategy runs.
+				if self.state.requesting_chunks.total_len() != 0 {
+					self.state.requesting_chunks = FuturesUndead::new();
+				}
+
+				gum::info!(
+					target: LOG_TARGET,
+					candidate_hash = ?self.params.candidate_hash,
+					"Starting `{}` strategy",
+					current_strategy.display_name(),
+				);
+
+				let res =
+					current_strategy.run(&mut self.state, &mut self.sender, &self.params).await;
+
+				match res {
+					Err(RecoveryError::Unavailable) =>
+						if self.strategies.front().is_some() {
+							gum::warn!(
```

Comment:
In practice it is not a serious issue if one strategy fails, I'd use `debug` here.